Add ActiveRecord::Relation#in_batches

siadat opened the pull request:
Hello everyone!

There have been many times when I needed to work with a relation in small chunks, e.g. deleting or updating a large number of rows while allowing the database to work on other tasks in between. This is especially useful if the number of affected rows is very large and we need to throttle our work in order to allow other processes to start/complete in between each iteration. It could also be used to process each slice using concurrent workers.

For example:

```
People.where('age > 21').each_slice do |relation|
  sleep 10 # other tasks
  relation.update_all(should_party: true)
end
```

Called without a block, it will return an `Enumerator`, e.g.:

```
Poeple.where('age < 18').each_slice.each(&:delete_all)
```

Other people have expressed interested in having such functionality as well. E.g., with a quick GitHub search I found #14465 and #13147. There have been other attempts to achieve the same goal, however they were backward-incompatible, not tested and not optimised.

I have written passing tests in a similar style to #find_in_batches. I have made sure that the number database queries is kept low (it is equal and in some cases less than that of #find_in_batches).

I have documented the method as well.

The method accepts a block, or returns an `Enumerator` if no block is give.

Please let me know if I need to make any modifications, or rename it so that I could get prepare it for getting merged.

PS. let me know what name you prefer for the method, e.g. #batches, #to_batches, #in_batches, #split, #each_slice, etc. The method is called #each_slice because it is dividing a single ActiveRecord::Relation object into multiple ActiveRecord::Relation objects, similar to the more familiar Array#each_slice, which divides an Enumerable into multiple Enumerable objects.

siadat committed:
Add ActiveRecord::Relation#each_slice

bogdan commented:
Here is an alternative implementation that modifies existing method and make it work with relation instead: https://github.com/rails/rails/pull/14465

siadat commented:
Thank you for your comment!

That implementation looks good as well. However, this one is for a different purpose, as it does not change the functionality of the existing methods. I believe it would be better if we do not change the functionality, and return types of existing methods. However, it could be change in the next major version, as @rafaelfranca has added it to 5.0.0 milestone. :+1:

One thing I am excited to note is that the new method I have submitted in the current pull-request is well tested and the number of queries in this method is less than half the number of queries in the implementation you mentioned and it is backward compatible as it does not change the functionality of existing methods.

bogdan commented:
@siadat I've explained why the number of queries in my PR is not reduced in a comment https://github.com/rails/rails/pull/14465#discussion_r10880138

I don't know myself which way is better. I personally want everyone to only use relation version of method instead of generic one.

number of queries optimisation can be tricky. I was thinking about it for a long time and didn't figure out an implementation (even if we drop backward compatibility) that will always have minimal number of queries. It depends on if the relation was loaded or not in the yield block: if it was then your solution is not optimal because you query the last element before calling yield.

So your solution actually doubles the number of queries if relation is loaded in block given. Which is kind of unavoidable.
 

I was thinking about new method with deprecation of the old one where you have something like this:

``` ruby
User.where(....).in_batches.each do |batch|
  batch.update_all(...)
end
User.where(....).in_batches(load: true).each do |batch|
  User.transaction do
    batch.each(&:touch)
  end
end
```

It have the following advantages:

* you optimise the number of queries yourself with `load: true` parameter
* easy migration: old batches method just gets deprecated and people can migrate smoothly.
* `in_batches` always returns an iterator allowing you to specify how you want to iterate it: `each` or `map`

matthewd commented:
> So your solution actually doubles the number of queries if relation is loaded in block given. Which is kind of unavoidable.

Can't we avoid that, for the simple case, by waiting to call `last` until after the yield?

matthewd commented:
Then again, doing a query before the yield (whether a `load` or a `last`) is probably the only way to address the biggest danger I can see here, which is overrun. To wit: consider what happens if the block calls `the_yielded_relation.where('random() < 0.5').each { .. }`... or any other, more plausible, partially-selective condition.

bogdan commented:
@matthewd thanks for the question. 

No, we can't: we need to know that relation is not empty before we yield. Otherwise we will yield `[]`  as last iteration. This is something I don't want to trade with. We need to ensure relation is not empty first.

Calling `size` before yield don't help as well because it is yet another query.

bogdan commented:
@matthewd maybe we should mark relation as readonly before passing it to yield. What do you think?

matthewd commented:
@bogdan I'm confused.. what would be the point? If you can't build on the supplied relation, what benefit does it offer over the current array?

bogdan commented:
@matthewd please see the initial description of `https://github.com/rails/rails/pull/14465#issue-30017062`.

dhh commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -143,6 +143,72 @@ def find_in_batches(begin_at: nil, end_at: nil, batch_size: 1000, start: nil)
+    #   end
+    #
+    # ==== Options
+    # * <tt>:batch_size</tt> - Specifies the size of the batch. Default to 1000.
```
To go with #each_slice, it should imo be `:slice_size`.

dhh commented:
On the API side, I think `#in_batches(of: 100).each do` is the best proposal I've seen. Then we can deprecate find_in_batches. It's definitely much nicer just to get a new relation you can work on with all the capabilities we have. If you guys can work together on that, it'd be great 👍

siadat committed:
Rename ActiveRecord::Relation#each_slice to ActiveRecord::Relation#in_batches

siadat committed:
Rename +batch_size+ option of ActiveRecord::Relation#in_batches to +of+

siadat commented:
Thank you @bogdan, @matthewd, and @dhh, I appreciate your feedback! :+1: 

I renamed the method and the size option as dhh suggested, so a usage example would be:

    People.in_batches(of: 100) do |relation|
      relation.update_all(on_the_dance_floor: true)
    end

Let me know if I need to do anything else to get this merged.

dhh commented:
Looks good on the API front. We should follow up with deprecations of find_each and find_in_batches.

siadat committed:
Deprecate #find_in_batch and #find_each

siadat commented:
I added deprecation warnings to `#find_each` and `#find_in_batches` method definitions, and wrapped their tests with `assert_deprecated`.

Let me know if I did anything wrong, or if there is anything else that needs to be done.

bogdan commented:
Ok lets start:

1. `in_batches` should have a `load` option that forces or not forces objects loading to optimise number of queries. Default for `load` I suppose should be `false`. The old behaviour of `find_in_batches` can be gained with `in_batches(load: true)`.

2. `find_each` should not be deprecated. Sometimes `find_each` is better then `in_batches` and sometimes isn't. `find_each` should be implemented using `in_baches(load: true)` so that it's behaviour remain the same. I can give examples why both methods should be in place if someone needs them

3. I am not sure that `in_batches do` is good idea. At least @dhh mentioned  `#in_batches.each` as the best variant and I agree with that.

4. Maybe `find_each` should be transformed to something like `with_batches.each` for consistency with (3)

dhh commented:
Bogdan, what's the case where `find_each` is an improvement over
`in_batches.each`? Also, can you explain the `load` option a bit more? I'm
not sure I'm getting the trade-offs. I agree that `in_batches.each` is
better than `in_batches do`.

On Fri, Jul 17, 2015 at 1:13 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> Ok lets start:
>
>    1.
>
>    in_batches should have a load option that forces or not forces objects
>    loading to optimise number of queries. Default for load I suppose
>    should be false. The old behaviour of find_in_batches can be gained
>    with in_batches(load: true).
>    2.
>
>    find_each should not be deprecated. Sometimes find_each is better then
>    in_batches and sometimes isn't. find_each should be implemented using in_baches(load:
>    true) so that it's behaviour remain the same. I can give examples why
>    both methods should be in place if someone needs them
>    3.
>
>    I am not sure that in_batches do is good idea. At least @dhh
>    <https://github.com/dhh> mentioned #in_batches.each as the best
>    variant and I agree with that.
>    4.
>
>    Maybe find_each should be transformed to something like
>    with_batches.each for consistency with (3)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20899#issuecomment-122248777>.
>

bogdan commented:
Ok All examples:



Here we want to update all records in batches because update whole table at once will lock table for writing:

``` ruby
User.in_batches.each do |user|
  user.update_all("disabled = true")
end
```

Here we want to touch individual record at ruby level because of some logic (e.g. chain touch), but we want to improve performance with transaction for each batch, so can't use `find_each`.
`load: true` here reduces number of queries because we call `relation.last` inside `in_batches` and it makes 2 queries per batch iteration while `load: true` will force loading and `relation.last` will be performed by ruby on loaded relation. 

``` ruby
User.in_batches(load: true).each do |relation|
  User.transaction do
    relation.each(&:touch)
  end
end
```

Here we don't want to process each batch individually and only care about iterating for each record. `find_each` used as a shortcut to `in_batches.each { |r| r.each { ... }}`:

``` ruby
data = User.find_each.map do |user|
  user.attributes.slice(:id, :email, :role).to_csv
end
```

Junior ruby developer would always use last one for all 3. Which is fine but slow. So `in_batches` is positioned as something advanced.

dhh commented:
For example #2, this should already work: `User.in_batches.load.each`. Don't think we need another option for it. Agree on Example #3. That's a nice convenience. First example should probably be `each do |batch|`, if used in docs, but yes.

dhh commented:
Sorry, for example #2, I meant `User.in_batches.each { |batch| transaction { batch.load.each(&:touch) } }`.

siadat commented:
@bogdan if I understand correctly, you are saying that when e.g. `load: true`, the method should return an Enumerator that yields arrays of ActiveRecord objects (as opposed to ActiveRecord::Relation objects). Because in that case `in_batches` method does not need to call `#last` anymore and could call `#to_a` instead. This would result in one less database query to those who want to call `#to_a` to iterate over each ActiveRecord object, like `find_each` does. If that is the case, then there might be a reason to keep `#find_each` for that purpose.

bogdan commented:
Example 2: `User.in_batches.each { |batch| transaction { batch.load.each(&:touch) } ` will not save you a query because you need to know if the batch is empty or not before yield it. From the other hand you can not always force loading because you don't know if it will be needed inside `yield`.

See https://github.com/rails/rails/pull/20899/files#diff-f0b4abc491880e438bf01f3f795e6ec1R213 
This is the query that `load: true` can avoid.

dhh commented:
Ah, I see. Fair enough!

On Fri, Jul 17, 2015 at 2:13 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> Example 2: User.in_batches.each { |batch| transaction {
> batch.load.each(&:touch) } will not save you a query because you need to
> know if the batch is empty or not before yield it. From the other hand you
> can not always force loading because you don't know if it will be needed
> inside yield.
>
> See
> https://github.com/rails/rails/pull/20899/files#diff-f0b4abc491880e438bf01f3f795e6ec1R213
> This is the query that load: true can avoid.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20899#issuecomment-122258009>.
>

siadat commented:
@bogdan :+1: Thank you!

I just realised that `#last` loads the relation because the limit is set. So we cannot avoid loading the relation if we are using `#last`.

To avoid loading, I think we could get the same results using `offset`. In which case, if `load: false` then only a COUNT query is executed on each iteration.

What is your thought about this:

    while true
      relation_yielded = relation.offset(offset)
      relation_yielded.load if load
      break if !relation_yielded.any?
      yield relation_yielded
      offset += of
    end

bogdan commented:
@siadat we can do it. But we need to ensure we don't break any test. In order to do that we need to implement `find_each` and `find_in_batches` using new `in_batches` method to ensure all it's concerns are covered in `in_batches`. 

In other words: all methods should use `in_batches` method as the most basic method and they should not change their behaviour. We can deprecate anything only after that.

siadat committed:
- #in_batches uses offset and counting, to iterate through the records
  without loading them. This is different from the previous approach,
  which worked with fetching the ID of the last item on each iteration.

- +:load+ option added to #in_batches reduce number of queries in case
  all records are iterated, by preloading the relation.

- #find_in_batches now uses #in_batches(load: true) to get all the
  records. This also affects #find_each, because #find_each calls
  #find_in_batches.

siadat commented:
@bogdan Thank you for helping out. I implemented `in_batches` using offset, and modified `find_in_batches` to use `in_batches` as well. I also updated the tests, and added two tests to ensure the `+:load` option works as expected.

I feel good about this, because even if a programmer iterates through all records while forgetting to set `load: true`, the unnecessary query executed is just a COUNT, which is far better than loading all the records on each iteration.

Let me know what you think.

matthewd commented:
I don't think this is viable as long as it's subject to [some variation of] the problem I described in https://github.com/rails/rails/pull/20899#issuecomment-121970395.

Using offset also appears to have an unpleasant behaviour if a record is inserted while we're iterating: potentially skipping it (as would be the case with limit) seems fair enough... double-processing some other record (as with offset) seems less so.

siadat commented:
@matthewd Thank you for your feedback. Could you explain the first problem a bit more. We are only overrunning a single COUNT query before stopping the iteration, i.e., the COUNT query that returns 0 when the offset is more than the number of matching records.

Regarding inserted records, we are ordering the query using the primary key as was the case in the original implementation. The original implementation would also process new records as well if it matched the conditions and was within the `+:end_at+` option. The only LIMITs applied were (and still are) the `+:end_at+` and the `+:batch_size+` (or the new `+:of+`) options. We are not double-processing new records if I am not mistaken. Again, please correct me if I am wrong.

egilburg commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -122,24 +132,82 @@ def find_in_batches(begin_at: nil, end_at: nil, batch_size: 1000, start: nil)
+      offset = 0
 
-        records = relation.where(table[primary_key].gt(primary_key_offset)).to_a
+      while true
```
`while true` can be replaced with `loop`

egilburg commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -122,24 +132,82 @@ def find_in_batches(begin_at: nil, end_at: nil, batch_size: 1000, start: nil)
+      while true
+        relation_yielded = relation.offset(offset)
+        relation_yielded.load if load
+        break if !relation_yielded.any?
```
`break if relation_yielded.none?`

siadat committed:
Use loop and none? as suggested by @egilburg

siadat commented:
Done. Thank you @egilburg!

bogdan commented:
@siadat 

1. ~~`in_batches` should have a `load` option~~
2. ~~`find_each` should be deprecated and implemented with `in_batches`~~
3. ~~`find_in_batches` should be deprecated and implemented with `in_batches`~~
4. `with_batches` is a new method with similar functionality as `find_each` but names of the options should be consistent with `in_batches`. (See Example 3 why)

@matthewd I don't see current implementation either makes the https://github.com/rails/rails/pull/20899#issuecomment-121970395 problem better or worse. This is yet another problem to solve some day.

siadat commented:
The query of the yielded relation could be changed by the programmer which might then result in unexpected matching rows. This is the case for any other method that returns a relation as well.

Would it make sense to ignore (or show a warning) if the yielded relation's query methods (#where, #order etc) are called?

siadat committed:
Add ActiveRecord::Relation#with_batches

It is a replacement for the original #find_each

siadat commented:
@bogdan Regarding `4` I added `with_batches`, which is used like `find_each`:

    User.with_batches(of: 100).each(&:touch)

bogdan commented:
@siadat `with_batches` looks good. 

> The query of the yielded relation could be changed by the programmer

Do you mean `#where!` and `#order!` relation modification methods? I feel I don't get it. Can you please make an example with code?

dhh commented:
Not liking with_batches. It's too close to in_batches and it doesn't
self-describe how they're different. find_each is clearly different.

On Saturday, July 18, 2015, Bogdan Gusiev <notifications@github.com> wrote:

> @siadat <https://github.com/siadat> with_batches looks good.
>
> The query of the yielded relation could be changed by the programmer
>
> Do you mean #where! and #order! relation modification methods? I feel I
> don't get it. Can you please make an example with code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20899#issuecomment-122529898>.
>

siadat commented:
@bogdan This would be an example of problematic usage:

    People.in_batches.each do |relation|
      relation.where(age: 21).update_all('age = age + 1')
    end

`relation.to_sql` would be something like this (as expected):

    SELECT * FROM people LIMIT 0, 1000

However, the added `where(age: 21)` changes everything:

    SELECT * FROM people WHERE age = 21 LIMIT 0, 1000

The programmer who wrote this would have probably meant to filter and narrow down the current batch. But the matching rows are changed completely because the added condition is applied before offset, not after it. Same thing happens even if we iterate each batch via `id > last_id_of_previous_batch` instead of offset.

One solution I could think of right now is this: we could load the primary key (e.g. id) of all matching rows, then create a new relation with them like this:

    loop do
      batch = relation.offset(offset)
      batch.load if load
      break if batch.none?
      ids = batch.pluck(:id)
      yield self.where(id: ids) # <- NOTE here
      offset += of
    end

This way the programmer would be free to do anything they want with the yielded relations. They could call `where`, `offset`, `order`, and all other query methods on it and the results would be exactly what is expected. If we do this, we could also revert back to the method of using last ids instead of offsets.

Let me know your thoughts about it.

siadat commented:
@dhh I agree. I would prefer a more descriptive name.

Maybe we could borrow from `Enumerable#flat_map` and call it `flat_each` or `in_flat_batches`!

bogdan commented:
@siadat I got it now. The solution for this problem definitely should not be part of this PR. We can not make it perfect just here.

sgrif commented:
Can we not just make `in_batches.flatten` be smart enough to do the right thing?

dhh commented:
Feels too cumbersome given how common this case is. Whatever we end up doing shouldn't feel like a regression vs #find_each. 



> On Jul 18, 2015, at 16:35, Sean Griffin <notifications@github.com> wrote:
> 
> Can we not just make in_batches.flatten be smart enough to do the right thing?
> 
> —
> Reply to this email directly or view it on GitHub.
>

matthewd commented:
@bogdan 

> The solution for this problem definitely should not be part of this PR. We can not make it perfect just here.

(Again, as much as I would like to land this API,) I'm not going to introduce a brand new shiny foot-gun and just hope it gets fixed later.

@sgrif any ideas? My only thought so far is that we load the current relation (to get the last ID), then add a `where(id: range)` clause and *copy the loaded data to that new relation*. :confused:

siadat commented:
Regarding my previous comment https://github.com/rails/rails/pull/20899#issuecomment-122541749, and as @bogdan advised, I submitted #20933 as an alternative implementation for `in_batches`.

Let me know what you think.

siadat committed:
Remove with_each and undeprecate find_each

bogdan commented:
I don't like current PR because it goes too far from current implementation. It has several points week points:

* We need have a suspicious benchmark that:
  * wasn't run on all databases
  * can not cover all possible database states and DB schemas
* We spawn a lot of queries for each batch

I think this PR have gone a wrong direction since we've changed implementation to remove `Relation#last` call.  The reasoning was right but the solution is wrong. Maybe it is not wrong some day, but not in this PR.

I propose to come back to `last` as a concept and fetch it using offset.
I think this is an algorithm we can use:

``` ruby
# last = A || B
last = records.offset(records.offset_value + batch_size - 1).first || records.offset(nil).last
```

* A - fetches last element from batch when it is full: has `batch_size` number of elements. 
* B - fetches last element for the last batch
